plan/statistics/statistics.go

```diff
@@ -510,6 +503,48 @@ type Builder struct {
 	PkOffset      int                        // PkOffset is the offset of primary key of integer type in the table.
 }
 
+func (b *Builder) buildMultiColumns(t *Table, offsets []int, baseOffset int, isSorted bool, done chan error) {
+	for i, offset := range offsets {
+		var err error
+		if isSorted {
+			err = t.build4SortedColumn(b.Sc, offset, b.IdxRecords[i+baseOffset], b.NumBuckets, false)
+		} else {
+			err = t.buildColumn(b.Sc, offset, b.ColumnSamples[i+baseOffset], b.NumBuckets)
+		}
+		if err != nil {
+			done <- err
+			return
+		}
+	}
+	done <- nil
+}
+
+const buildColumnConcurrency = 8
+
+func (b *Builder) splitAndConcurrentBuild(t *Table, offsets []int, isSorted bool) error {
+	offsetCnt := len(offsets)
+	groupSize := (offsetCnt + buildColumnConcurrency - 1) / buildColumnConcurrency
+	splittedOffsets := make([][]int, 0, buildColumnConcurrency)
+	for i := 0; i < offsetCnt; i += groupSize {
+		end := i + groupSize
+		if end > offsetCnt {
+			end = offsetCnt
+		}
+		splittedOffsets = append(splittedOffsets, offsets[i:end])
+	}
+	doneCh := make(chan error, len(splittedOffsets))
+	for i, offsets := range splittedOffsets {
+		go b.buildMultiColumns(t, offsets, i*groupSize, isSorted, doneCh)
+	}
+	for range splittedOffsets {
+		err := <-doneCh
```

Comment:
I was following the logic at https://github.com/pingcap/tidb/blob/master/domain/domain.go#L103.